### PR TITLE
[YC W23] adding support for Claude, Replicate-Llama2, Cohere

### PR DESCRIPTION
**Pull request by krrishdholakia**
Hey guys, 

Noticed y'all are using multiple LLM Providers - used `litellm` (simple library to standardize LLM API Calls - https://github.com/BerriAI/litellm)  to add 3 new providers: Claude, Replicate, Cohere

I specified them to be `claude-instant-1`, `llama2..`, `command-nightly` in the classes but let me know if there's any changes you'd like me to make. 

Would love to know if this helps.

### CODE REVIEW

**Comment by krrishdholakia** [comment]
should also have palm2 ready in a few days

File: src/llm_vm/onsite_llm.py [hunk 1]
```diff
@@ -467,3 +467,110 @@ def finetune(self, dataset, optimizer, c_id):
...
+        """
+        cur_prompt = [{'role': "system", 'content' : prompt}]
+        ans = completion(
+            messages=cur_prompt,
+            model="claude-instant-1", # could also do claude-2
+            **kwargs)
+        return ans['choices'][0]['message']['content']
+
+    def finetune(self, dataset, optimizer, c_id):
+        print("fine tuning isn't supported by OpenAI on this model")
```
Comment (cartazio) [inline src/llm_vm/onsite_llm.py:503]: this seems wrong :) but easy to fix

File: src/llm_vm/onsite_llm.py [hunk 2]
```diff
@@ -467,3 +467,110 @@ def finetune(self, dataset, optimizer, c_id):
         # optimizer.storage.set_training_in_progress(c_id, False)
         # if old_model is not None:
         #     openai.Model.delete(old_model)
+@RegisterModelClass("claude")
+class Claude:
+    """
+    This is a class for openAI's gpt-3.5-turbo LLM
```
Comment (cartazio) [inline src/llm_vm/onsite_llm.py:473]: fix this typo :)

File: src/llm_vm/onsite_llm.py [hunk 3]
```diff
@@ -467,3 +467,110 @@ def finetune(self, dataset, optimizer, c_id):
         # optimizer.storage.set_training_in_progress(c_id, False)
         # if old_model is not None:
         #     openai.Model.delete(old_model)
+@RegisterModelClass("claude")
+class Claude:
```
Comment (cartazio) [inline src/llm_vm/onsite_llm.py:471]: where does this instance get its state for doing the claude api

File: src/llm_vm/onsite_llm.py [hunk 4]
```diff
@@ -467,3 +467,110 @@ def finetune(self, dataset, optimizer, c_id):
...
+        return ans['choices'][0]['message']['content']
+
+    def finetune(self, dataset, optimizer, c_id):
+        print("fine tuning isn't supported by OpenAI on this model")
+        exit()
+
+@RegisterModelClass("llama2-replicate")
+class Llama2Replicate:
+    """
+    This is a class for openAI's gpt-3.5-turbo LLM
```
Comment (cartazio) [inline src/llm_vm/onsite_llm.py:509]: fix this

**Comment by krrishdholakia** [comment]
Thanks for the feedback - will make the changes and revert

**Comment by cartazio** [comment]
I think we're generally requiring any integration aside from openai (claude is private beta for its api right?) to be runnable locally. Additionally, we're generally having the following (implicit) expectations for models we include.

1) if its a family of models, the default for a "by name" selection should be the smallest one in a given model family. This is to make contributing and experimentation in the spirit of open source accessible to everyone. Defaults that *OOM* when you try them aren't great! You'll see thats what we do with the other examples we include. 

2) I think for any locally runnable LLMs, they need to provide hf-transformers compatible interfaces. This means, at least conceptually: 
 a) they need to be able to provide the generic completions api, but also expose the probability vector of choices over next tokens, and also the related requirement of exposing a full tokenizer interface. 
 b) some sort of finetuning story needs to be provided (our current wrappers are lack luster here, but they are a stab in the right direction)

**Comment by krrishdholakia** [comment]
hmm, how do you get log probs for openai? i believe even chatcompletion doesn't provide that

**Comment by cartazio** [comment]
Indeed! Openai actually is a feature regression in that respect! It just
happens to be a model that everyone reaches for, and works pretty well as a
“teacher” in student teacher configurations.

So the architectural faux pas is we conflate student and teacher model
apis. Which is probably something we should fix eventually.  And I think
that’s also a design nuance we should document better

**Pull request closed without merging**